flang/lib/Lower/IntervalSet.h

```diff
@@ -0,0 +1,98 @@
+//===-- IntervalSet.h -------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FORTRAN_LOWER_INTERVALSET_H
+#define FORTRAN_LOWER_INTERVALSET_H
+
+#include <cassert>
+#include <map>
+
+namespace Fortran::lower {
+
+//===----------------------------------------------------------------------===//
+// Interval set
+//===----------------------------------------------------------------------===//
+
+/// Interval set to keep track of intervals, merging them when they overlap one
+/// another. Used to refine ranges of offsets.
+struct IntervalSet {
+  using MAP = std::map<std::size_t, std::size_t>;
+  using Iterator = MAP::const_iterator;
+
+  // Handles the merging of overlapping intervals correctly, efficiently.
+  void merge(std::size_t lo, std::size_t up) {
+    assert(lo <= up);
+    if (empty()) {
+      m.insert({lo, up});
+      return;
+    }
+    auto i = m.lower_bound(lo);
+    // i->first >= lo
+    if (i == begin()) {
+      if (up < i->first) {
+        // [lo..up] < i->first
+        m.insert({lo, up});
+      } else {
+        // up >= i->first
+        if (i->second > up)
+          up = i->second;
+        m.erase(i);
+        // merge i with [lo..max(up,i->second)]
+        m.insert({lo, up});
+      }
+    } else {
+      if (i == end() || i->first > lo)
+        i = std::prev(i);
+      // i->first <= lo
```

Comment:
This must be `i->first < lo` if it could be equivalent m.lower_bound(lo) would return this iterator and not the next one.

Reply:
The comment is correct.
`i >= lo` at the top of the `else` block. The `if` backs up the iterator `i` only if `i > lo`. So if `i == lo`, the iterator is unchanged. Otherwise the iterator points to the previous interval, where it is guaranteed that `i < lo`. Combining, we have `i <= lo`.

Reply:
I see what you're saying.